### ISSUE TITLE
Width measurement

### ISSUE THREAD

**Issue by raphlinus** [google/xi-editor#565]
This is the tracking issue for width measurement in xi-editor, and something of a design document and list of concerns.

## Introduction

A number of editing operations require some concept of width. Probably the most important is word wrap (see #58 for more issues regarding improving word wrapping). Currently, utf-8 code unit count is used as a stand-in for all width operations, but results are poor, as these only loosely correlate (issue #501 describes the consequences in more detail, giving examples).

Other operations that require width include:

* Computing a location within a line with similar horizontal positioning, for up/down 1 line and similar movements. (This is represented in `SelRegion::horiz`)

* Computing the entire-document width, as one alternate for horizontal scrolling (google/xi-mac#145, #479).

* Expansion of tab stops. (It's not obvious yet whether these will be expanded in the front-end or core, see #467).

* Column indicator in status UI.

## Statement of the problem

For the purposes of word wrapping, the fundamental problem is to compute the width of a string, with merged attributes, in the current font and font size choice selected in the front-end. The task of word wrapping is then defined as choosing, for each line, the last line break opportunity that satisfies the constraint that the width of the line (ie substring from the previous break to this break) is less than or equal to the wrap width. (As a detail, trailing space is not included in this width calculation)

In general, width measurement is not a homomorphism, ie for strings A and B, width(A + B) is not width(A) + width(B). However, for the purposes of efficiency, we will assume in some cases that it's close. In particular, if a line break opportunity exists between A and B, then we can assume the widths are independent. In some cases, this might lead to incorrect line breaks, but the amount of mismeasurement is likely to be modest.

Horizontal positioning within a line is much trickier to solve exactly, as the positioning of carets within a word is subject to ligatures (fonts sometimes have a ligature caret table, but this is something of an advanced feature). In addition, it requires solving the inverse problem (resolving a width to an offset within the string). As a simplification and performance improvement, I suggest the approximation of assuming that each grapheme cluster within the word has the same width. The consequence of mismeasurement should be pretty modest; it will lead to the cursor "wiggling" left and right a bit more on repeated up/down arrow key requests. Note that in the general (non-monospace) case there is already nonzero wiggle as the position needs to be rounded to the nearest grapheme cluster boundary in any case.

Note that the front-end needs to solve the precise version of this problem to display a caret, and to resolve a click or touch gesture to a position within the string. These cases require precision, and I do not propose using an approximation; it is strictly for computations within the core.

## Performence: RPC, caching

The core needs width measurement, but the front-end is the most natural place for the calls to the underlying text engine, as the front-end needs to do text layout as part of its text rendering.

Even without the core/front-end separation, text measurement tends to be a slow operation. Every word (more precisely, every substring between consecutive break opportunities) needs to be measured, so the number of measurement opporunities is potentially large. One blocking RPC per word would have unacceptable performance. We can and should employ a number of techniques to improve performance.

First and foremost is a cache. It's likely cache effectiveness is high, and almost guaranteed in the case of re-wrapping a paragraph being edited (or when the wrap width is changed, as in the case of resizing a window). Android uses a simple heuristic for granularity of its word cache: whitespace, or a boundary between to CJK ideographs.

Second, we can use a fast-path for the simple case of a monospace font. This will be particularly effective for large files such as logs, but we should be careful not to rely on it too heavily, as it's not effective for complex scripts, emoji, or even non-monospace fonts for ASCII.

Third, to reduce RPC overhead, we can apply a number of techniques. Without too much complexity we can batch requests (so each RPC is for an array of attributed words). Adding more complexity, we can pipeline, but the benefits are likely not strong.

Lastly, we could avoid RPC overhead by bundling the core and front-end into the same process and making width measurement a procedure call, see #439 for the prerequisite issue. I believe this should be an optional performance optimization rather than a requirement (for isolation/robustness reasons and to make it easier to stand up a new front-end), but after gaining empirical experience we can decide that RPC overhead is minimal, unacceptable, or somewhere in between.

## Other details

The RPC call should be based on the style id (`set-style` in the core to front-end update protocol). If we do libxi, then we probably need to move `set-style` calls to procedure calls also, to prevent the possibility of the width measurement request being reordered before the dependent style definition by RPC transport asynchrony.

## "Columns"

It is common UI in editors to show a "line/column" indication. For ASCII (or even ISO 8859-1), the units are clear, but for complex scripts and emoji, it's not clear how columns should be counted.

Sublime Text counts Unicode codepoints. While it's a clear metric, it's not always satisfying, as any grapheme made from multiple codepoints counts as multiple columns.

We don't currently have this feature, but it is likely to be requested. One approach is to use measured width divided by the width of a "nominal" character.

**Comment by vlovich** [comment]
On mobile so this is more a stream of consciousness then anything
particularly deeply considered so apologies ahead of time if my questions
are nonsensical.

For the up/down movement/selection, isn’t that a front-end problem?  Maybe
plugins might need it but does core actually need it?

Document width - this seems like something that can just be computed in the
front-end and just communicated to the backend when it changes (again -
unclear to me that core itself actually needs it vs computing itself for
display) if the backend needs the info.  It also changes infrequently and
should be cheap to recompute as the user types since you only need the
longest width + width of lines being modified.

Expansions of tab stops.  I agree this is less clear for core. My first
instinct is that it’s a front end problem but I could see plugins needing
that info.

Column indicator - this seems again like a front-end problem.  Even if a
plug-in/core wants to access by column count the front-end can just accept
the column count and convert it.  Also, is it actually expensive to
compute?  You need to render everything anyway so the width should already
be available for everything on the current line anyway.  Super long lines
obviously pose a problem since there’s a natural cap to how much you
compute for rendering but I feel like that’s a special case optimization.
Now jumping the cursor to an arbitrary line+col might be expensive but my
gut instinct is that it won’t be too bad since that’s an operation wholly
within the front end.

So based on those intial thoughts my impression is that for now it seems
like the use cases for width are either in the front-end or are a much
simpler subset that aren’t initiated by core (or pushed by frontend as
stats change).  I’m sure you’ve thought about this more so I’d love to
learn more to correct my understanding.

**Comment by raphlinus** [comment]
@vlovich: thanks for the questions, it very much helps clarify my own thinking. I'll try to address them below.

**Up/down movements in front-end:** This one is interesting because it's statistically _very_ likely that all the information needed to compute the movement is in the FE, but it's not true in the general case. Consider: load a large file; select all; split selection into lines, right arrow, down.

That said, I'm turning over in my head whether we want to do an approximation in the core, or just open up an RPC to the FE for it. The latter is more RPC surface, but it also means there is only one implementation of the offset-to-horiz and horiz-to-offset logic.

**Document width:** the main challenge here is on document load. However, it doesn't need to be particularly fast, it's fine if the width calculation happens asynchronously after the document loads. Then, the UX is that the h scrollbar adjusts after initial load.

**Tab stops:** my gut feeling is that this is FE not core, also. Among other things, the assumption that a tab can be expanded to an integral number of spaces breaks down in the face of complex scripts and non-monospace fonts.

**Column indicator:** I'm still struggling with this one. The reason why it's not obviously a FE concern is that it's computed from the start of the paragraph, not the start of the visual line. If the paragraph is very long, then it may not be fully present in the FE's line cache. On the other hand, thinking about this, presenting a width-based calculation is going to be non-intuitive for non-monospaced fonts. I could be convinced that just counting codepoints is Good Enough, and it wouldn't be all that hard to implement.

I mostly wanted to capture these other potential uses of width measurement; some of them may turn out not to be relevant, or at least very important. The primary one, and the one that's by far most performance-sensitive, is word wrapping. Re-wrapping a very large document when the window width changes involves potentially a lot of computation. There are async tricks that can be done to hide much of it, but that ratchets up overall complexity.

Hope this helps answer some of your questions.

**Comment by raphlinus** [comment]
I should probably add to this the concrete proposal for a new RPC in the core->front-end protocol.

```
measure_width [{"style": 42, "strings": ["abc", "def"]}]
```

The return value is an array of array of (float) width values, measured in CSS px units, or the closest analog available (on macOS these are called "points", not to be confused with the [typography]( https://en.wikipedia.org/wiki/Point_(typography)) meaning).

In addition, the front-end should be able to send an `invalidate_widths` notification to the core, which would trigger a re-wrap of the document. Another, similar notification (outside of the scope of this issue) would update the viewport width of a view, similarly triggering re-wrap but not affecting the validity of width measurements.

**Comment by vlovich** [comment]
I’m not sure I fully grasp the up/down movement in the example.  Isn’t the
cursor position some offset into a line based on glyph count rather than
some width calculation?  I’m not quite sure I understand why the core would
need graphical width info in this case.

For document load I think the UX is going to be the same (whether you defer
load or reflow the scroll bar) regardless of where you do it since the cost
seems to me like it should be the same in the backend and frontend.

Tab stops I agree but again I’m not sure I understand why this isn’t a
rendering issue but at the same time I’m surprised by the current expand
tabs + Unicode bug so I’m sure I’m missing something about the actual
implementation vs the simplified “ascii” based thinking I’m in.

For the column indicator, I think counting some Unicode unit will be more
consistent.  Am I correct in understanding the point you’re trying to get
to is to have the same column number across lines at the same x position
even for non monospace fonts?  If so I think that’s probably
counterintuitive for users who will probably expect column + 1 for hitting
the cursor right.  Counting the Unicode units seems like the more
consistent approach to me as an end user.  It will obviously require some
more thought if we were to support the rich tab stop features like Word has
but that again feels like a more FE concern the backend wouldn’t be
involved in (at least that’s my gut instinct).

For word wrapping, why does that need to be done in core?  Even if it did,
it seems to me like the frontend can just tell the backend where to put the
line breaks for the current view (or probably a bit before and after)
rather than doing it upfront for the entire document, no?

**Comment by mqzry** [comment]
>  Am I correct in understanding the point you’re trying to get
to is to have the same column number across lines at the same x position
even for non monospace fonts. If so I think that’s probably
counterintuitive for users who will probably expect column + 1 for hitting
the cursor right.

Why would 'column + 1 on cursor right' not work as users expect then?

**Comment by raphlinus** [comment]
@vlovich Right, I should clarify that the unit of measurement for "horiz" in the case of vertical movement is a choice. I think the best choice is for it to be measured width, but other choices are valid. I think if you test editors that handle proportional fonts, you'll find that most of them use measured width for this. A good way to test it is to put "mmii" on one line, then "iimm" on the next, position cursor at "m|mii", then down. Most editors will end up at "ii|mm", I think.

Also, whether to word wrap in core or front-end is also a choice, and both could be made to work. One issue with doing it in the FE is that if you have a very long paragraph, all of it would need to be present in the FE. I think an even more compelling argument is that re-wrapping a paragraph when you're typing is a fairly tricky delta computation (I only want to repaint the lines that have actually changed), and I want that in core. By wrapping in core, the core->FE update can have line granularity, and the FE can have pretty simple logic.

Wrapping in FE vs wrapping in core has a user-visible impact on scrolling. By doing it in core, we always know the exact physical height of the document, and the correspondence document position

Bottom line, I've made the decision, wrapping is in core, and I'm not strongly moved to reconsider it.

I'm going to withdraw my proposal to make column indicator width-based, and should probably create another issue to explore what units should be used.

I didn't understand the point about document load. A major goal is to make the FE only have to deal with the portion of the document that's visible. I'm wondering whether I've miscommunicated some of my basic assumptions - if so, it might be worth writing them more clearly.

**Comment by vlovich** [comment]
Hmm... regarding the movement, I'm still not sure I understand.  I agree that measured width is probably correct as it gives correct behaviour regardless of fonts.  I think my point is that wouldn't the FE be able to do the translation of measured width -> abstract backend cursor position -> notify BE of the desired location?  I'm not sure I get why the BE needs to support cursor movement in terms of measured width only to have to need to query the FE for the width instead of doing the measured width directly.

RE line-wrapping, I understand your point now much better & I 100% agree with your decision.  That being said, I'm not 100% convinced it needs measured width.  I think it ties into how we do columns.  For programming use-cases, width is typically defined in terms of # of columns & it feels natural this is independent of font width.  For WYSIWG document editors I agree line wrapping based on measured width makes more sense since the line-wrapping is page width.  For regular text editing, it's probably also measured width based if you wrap to the window.  I think it might help to clarify Xi's priorities for code editing vs general text editing vs WYSIWG editors, but having thought about this more I understand why a measured width RPC protocol would be needed so we could support wrapping to window.

For document load I think I just didn't understand what you were saying.  Am I correct in saying that you want to compute the hscroll width based on measured width instead of glyph counts?  In that case I understand the need for the BE to request the FE to measure the width.
